web-app/HeimScripts/heatmap/run.R

```diff
@@ -1,187 +1,254 @@
-    sds <- sd(df[,3],na.rm = T) #For just one sample we take mean over all genes (whole column), not per gene
+    variance <-
+      1.0  # We cannot get any significance measure for just
+           # one sample
+    mean <-
+      mean(df[,3], na.rm = T)  # For just one sample we take mean
+                               # over all genes (whole column), not per gene
+    sds <- sd(df[,3],na.rm = T)  # For just one sample we take mean
+                                 # over all genes (whole column), not per gene
     df["MEAN"] <- rep(mean,nrow(df))
     df["SD"] <- rep(sds,nrow(df))
     df["SIGNIFICANCE"] <- rep(variance, nrow(df))
   }
-  df <- df[1:min(max_rows,nrow(df)),]
-  fields <- buildFields(df)
-  extraFields <- buildExtraFields(fields)
-  probes <- na.omit(df[,1])
-  geneSymbols <- df[,2][1:length(probes)]#unique(fields["GENESYMBOL"])[,1] #[,1] in order to get a vector, otherwise we get a dataframe
-  patientIDs <-unique(fields["PATIENTID"])[,1]
+  return(df)
+}
 
-  significanceValues <- unique(fields["SIGNIFICANCE"])[,1]
-  features <- unique(extraFields["FEATURE"])[,1]
-  jsn <- list("fields"=fields, "geneSymbols"=geneSymbols,
-                              "patientIDs"=patientIDs,
-                              "probes"=probes,
-                              "significanceValues"=significanceValues,
-                              "features"=features,
-                              "extraFields"=extraFields
-                               )
-  writeDataForZip(df, patientIDs) # for later zip generation
-  writeRunParams(max_rows, sorting)
-  df <- cleanUp(df) # temporary stats like SD and MEAN need to be removed for clustering to work
-  measurements <- df[,3:ncol(df)]
-  if (nrow(measurements) > 1 && ncol(measurements) > 1 ){# cannot cluster matrix which is less than 2x2
-    measurements <- toZscores(measurements)
-    jsn <- addClusteringOutput(jsn, measurements) #
+parseInput <- function() {
+  if (exists("preprocessed")) {
+    df <- preprocessed
   }
-  jsn <- toJSON(jsn,
-                pretty = TRUE)
-  write(jsn,file = "heatmap.json") # json file be served the same way like any other file would - get name via /status call and then /download
-
-  msgs <- c("Finished successfuly")
-  if (exists("errors")){
-    msgs <- errors
+  else {
+    df <- mergeFetchedData(loaded_variables)
   }
-  list(messages=msgs) # main function in every R script has to return a list (so a data.frame will also do)
+  return(df)
 }
 
-toZscores <- function(measurements){
+toZscores <- function(measurements) {
+  print(colnames(measurements))
   measurements <- scale(t(measurements))
   t(measurements)
 }
 
-cleanUp <- function(df){
+cleanUp <- function(df) {
   df["MEAN"] <- NULL
-  df["SD" ]  <- NULL
+  df["SD"]  <- NULL
   df["SIGNIFICANCE"] <- NULL
   df
 }
 
-verifyInput <- function(max_rows, sorting){
-  if(max_rows <= 0 ){
+verifyInput <- function(max_rows, sorting) {
+  if (max_rows <= 0) {
```

Comment:
No. Heatmap can be created just fine for single row - but the clustering will not be done - because it is not possible.